spec/compiler/semantic/proc_spec.cr

```diff
@@ -881,15 +881,15 @@ describe "Semantic: proc" do
       )) { proc_of(int32) }
   end
 
-  it "merges Proc that returns Nil with another one that returns something else (#3655)" do
+  it "*doesn't* merge Proc that returns Nil with another one that returns something else (#3655) (this was reverted)" do
     assert_type(%(
       a = ->(x : Int32) { 1 }
       b = ->(x : Int32) { nil }
       a || b
-      )) { proc_of(int32, nil_type) }
+      )) { union_of proc_of(int32, int32), proc_of(int32, nil_type) }
```

Comment:
@bcardiff This is similar to the spec that you want. In a previous PR I made this always be `Proc(Int32, Nil)`, and this PR reverts that. But I will add a test that `call` returns the union.

Reply:
Cool, thanks. Now we need a rebase on master due to conflicts

Reply:
Done!